registry/dynamodb.go

```diff
@@ -0,0 +1,447 @@
+	if im.cacheInterval > 0 {
+		im.recordsCache = endpoints
+		im.recordsCacheRefreshTime = time.Now()
+	}
+
+	return endpoints, nil
+}
+
+// MissingRecords returns nil because there are no missing records for DynamoDB registry.
+func (im *DynamoDBRegistry) MissingRecords() []*endpoint.Endpoint {
+	return nil
+}
+
+// ApplyChanges updates the DNS provider and DynamoDB table with the changes.
+func (im *DynamoDBRegistry) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
+	filteredChanges := &plan.Changes{
+		Create:    changes.Create,
+		UpdateNew: filterOwnedRecords(im.ownerID, changes.UpdateNew),
+		UpdateOld: filterOwnedRecords(im.ownerID, changes.UpdateOld),
+		Delete:    filterOwnedRecords(im.ownerID, changes.Delete),
+	}
+
+	statements := make([]*dynamodb.BatchStatementRequest, 0, len(filteredChanges.Create)+len(filteredChanges.UpdateNew))
+	for _, r := range filteredChanges.Create {
+		if r.Labels == nil {
+			r.Labels = make(map[string]string)
+		}
+		r.Labels[endpoint.OwnerLabelKey] = im.ownerID
+
+		key := r.Key()
+		oldLabels := im.labels[key]
+		if oldLabels == nil {
+			statements = append(statements, &dynamodb.BatchStatementRequest{
+				Statement: aws.String(fmt.Sprintf("INSERT INTO %q VALUE {'k':?, 'o':?, 'l':?}", im.table)),
+				Parameters: []*dynamodb.AttributeValue{
+					toDynamoKey(key),
+					{S: aws.String(im.ownerID)},
+					toDynamoLabels(r.Labels),
+				},
+				ConsistentRead: aws.Bool(true),
+			})
+		} else {
+			im.orphanedLabels.Delete(key)
+			statements = im.appendUpdate(statements, key, oldLabels, r.Labels)
+		}
+
+		im.labels[key] = r.Labels
+		if im.cacheInterval > 0 {
+			im.addToCache(r)
+		}
+	}
+
+	for _, r := range filteredChanges.Delete {
+		delete(im.labels, r.Key())
+		if im.cacheInterval > 0 {
+			im.removeFromCache(r)
+		}
+	}
+
+	oldLabels := make(map[endpoint.EndpointKey]endpoint.Labels, len(filteredChanges.UpdateOld))
+	for _, r := range filteredChanges.UpdateOld {
+		oldLabels[r.Key()] = r.Labels
+
+		// remove old version of record from cache
+		if im.cacheInterval > 0 {
+			im.removeFromCache(r)
+		}
+	}
+
+	for _, r := range filteredChanges.UpdateNew {
+		key := r.Key()
+		statements = im.appendUpdate(statements, key, oldLabels[key], r.Labels)
+
+		// add new version of record to caches
+		im.labels[key] = r.Labels
+		if im.cacheInterval > 0 {
+			im.addToCache(r)
+		}
+	}
+
+	err := im.executeStatements(ctx, statements, func(request *dynamodb.BatchStatementRequest, response *dynamodb.BatchStatementResponse) error {
+		var context string
+		if strings.HasPrefix(*request.Statement, "INSERT") {
+			if aws.StringValue(response.Error.Code) == "DuplicateItem" {
+				// We lost a race with a different owner or another owner has an orphaned ownership record.
+				key := fromDynamoKey(request.Parameters[0])
+				for i, endpoint := range filteredChanges.Create {
+					if endpoint.Key() == key {
+						log.Warnf("Skipping endpoint %v because owner does not match", endpoint)
```

Comment:
Warnings only for deprecations please. I think it's more an Info log, than Error.